bson/src/main/org/bson/codecs/DocumentCodec.java

```diff
@@ -216,10 +229,29 @@ private Object readValue(final BsonReader reader, final DecoderContext decoderCo
             return null;
         } else if (bsonType == BsonType.ARRAY) {
             return readList(reader, decoderContext);
-        } else if (bsonType == BsonType.BINARY && BsonBinarySubType.isUuid(reader.peekBinarySubType()) && reader.peekBinarySize() == 16) {
-            return registry.get(UUID.class).decode(reader, decoderContext);
+        } else {
+            Codec<?> codec = bsonTypeCodecMap.get(bsonType);
+            
+            if (bsonType == BsonType.BINARY && reader.peekBinarySize() == 16) {
+                switch (reader.peekBinarySubType()) {
+                    case 3:
+                        if (uuidRepresentation == UuidRepresentation.JAVA_LEGACY
+                                || uuidRepresentation == UuidRepresentation.C_SHARP_LEGACY
+                                || uuidRepresentation == UuidRepresentation.PYTHON_LEGACY) {
+                            codec = registry.get(UUID.class);
```

Comment:
I'll have to think about this one.  There's an implicit assumption here that the Codec<UUID> has the same UuidRepresentation as this codec, so this is a bit brittle.  Not sure if it will be a problem in practice though.

Reply:
So in 3.12 can this change in DocumentCodec be reverted?

Reply:
Interesting, So what does passing in UuidRepresentation provide here, if the registry can ignore it.

Reply:
Not sure what you mean @rozza.  Passing it into `DocumentCodec`?

Reply:
Yes, its no guarantee that the correct uuid representation will actually be used.

Reply:
That's right.  We could add a `getUuuidRepresentation` method to `UuidRepresentationOverridingCodec` and do an instanceof check on the codec to ensure it has the same representation, but there is no guarantee that the Codec actually implements that interface (ours would, but a user _could_ provide their own codec that does not.